### ISSUE TITLE
High CPU usage, seemingly related to watcher notifications about folders

### ISSUE THREAD

**Issue by JonasKvarnstrom** [canton7/SyncTrayzor#401]
For various reasons, I want to synchronize a Windows user home folder from a laptop to another computer.  (I'm perfectly aware that sync is not backup, and I have some specific reasons for wanting to do this anyway.)

Unfortunately this has led to comparatively high CPU usage for SyncTrayzor / Syncthing.  Even when the computer is left idle, and most files that are regularly modified are ignored, the sync process uses an average of about 15%-20% of one core (i7-4600u at 2.70 GHz).  By "average" I mean that there are bursts where it uses a full core and idle intervals inbetween.

Turning on some of the Syncthing debug log options shows that changes to individual files are handled quickly: Going from "Walk..." to "Walk progress done..." typically takes 10 to 20 milliseconds.  But then the file watcher also quite often reports changes to *folders* such as the home folder itself:

> [YXMEM] 2017/08/26 20:15:15.006664 walk.go:102: DEBUG: Walk \\?\C:\Users\jonkv82 [] 131072 &{...}

I don't know what causes this (what actually happened to the jonkv82 folder to cause a notification).  But after this message, it usually takes up to 18-20 seconds of high CPU usage until we get the corresponding "Walk progress done".  This is consistent with walking the entire subtree of about 140,000 files due to what I assume is some kind of change in the *folder* itself.  This seems to happen rather regularly once every 15 minutes, plus some "random" times in between.

Similarly, for AppData:

> [YXMEM] 2017/08/26 20:19:09.323512 walk.go:102: DEBUG: Walk \\?\C:\Users\jonkv82 [AppData] 131072 &{...}

This typically takes about 11-12 seconds for 90,000 files, and can happen more often.

I took a brief look at the Syncthing source in walk.go where these messages are generated.  I've never used Go before so I could be mistaken, but it seems that Filesystem.Walk does walk the entire directory tree beneath my home folder or beneath AppData.  I assume this is in fact not necessary because if there were other modified files beneath these folders, the file system watcher would specifically notify SyncTrayzor about each file without the need for a full directory traversal?

I'd be happy to provide any additional information that might be useful.

**Comment by canton7** [comment]
If we see several changes to files in a folder, then we tell Syncthing to rescan the folder. This has the upside that we don't constantly bombard it with requests if there are lots of changes, but the downside that if you have a folder with many many files, Syncthing will end up doing unnecessary scanning. There are other cases which will cause a scan of the folder, such as a file being deleted (or moved).

Syncing your whole home directory is normally a bad idea. Things like ntuser.dat, and all of your temporary files, live in your home directory. This includes application-specific data, and applications normally don't like that being changed under them. It also includes things like the SyncTrayzor log file, which will of course log every time SyncTrayzor records a change to the log file....

I'd instead recommend syncing specific directories (you can use ignores to ignore everything except some specific folders). A less preferable option is to ignore files which shouldn't be synced (ntuser.dat, etc), and all temporary frequently-changing files.

**Comment by JonasKvarnstrom** [comment]
I see. 

To clarify my use case, I currently use Crashplan to back up my laptop to several other computers with a 5-minute interval, but they are cancelling their service for home users and I need an alternative.  I'm using BorgBackup for many other purposes, but its Windows version is not well tested and it needs to rescan the entire folder hierarchy every time it runs, which isn't well suited for a more "continuous" form of backup on a laptop where you might save a file and then put the laptop to sleep a minute later.  Many other backup programs seem to have the same weakness relative to Crashplan.

Therefore I thought it would be an interesting experiment to (1) run SyncTrayzor to immediately mirror everything of importance to a stationary computer that's permanently on, and then (2) run BorgBackup to back up the mirrored folder on the *stationary* computer at regular intervals, where I don't have to worry about power usage or about the computer shutting down before the entire folder is scanned and all files are backed up.  Kind of a weird setup but the experiment actually seems to be working pretty well so far, apart from the CPU usage.  An additional advantage over many backup programs is that Syncthing takes care of finding my computer despite its dynamic IP address, opening ports in the firewall, and so on.

I have set up various exclusions so that temporary files and other "non-restoreable" files are not mirrored.  I have also ensured the folder is "send only" on the laptop because the intention is to never synchronize anything back to the home folder.  In case of a disaster I'll selectively restore files from the BorgBackup archives.

Setting up individual folders would perhaps be a workaround but as far as I can tell this has some drawbacks.  First, to actually decrease the number of files scanned, I'd have to set up a rather large number of individually synchronized folders -- at least one for each folder under AppData\Roaming, and then there are plenty of other folders of interest as well.  Second, I'd have to add new folders every time something new appears.  For example, I use IntelliJ IDEA which sets up a new config folder for each new major version, three times per year.  I'd probably forget to do this at some point.  Having one main folder with explicit exclusions would work better for me in this specific use case, and if I happen to include something that shouldn't be restored, it won't really hurt that it's part of the backup archives.

Thanks for the information that SyncTrayzor consolidates multiple calls into a single folder update!  I have excluded a number of frequently changing files and thought I had basically caught them all, but I found them by analyzing the Syncthing log entries which won't work if SyncTrayzor has already consolidated entries before Syncthing can log them.  I will look for other ways of finding that information and maybe this will reduce the scan frequency.  I'll comment here as soon as I've tested (probably tomorrow).

**Comment by JonasKvarnstrom** [comment]
I'm trying to understand exactly how the notification logic works relative to .stignore.  The main github page for SyncTrayzor says that "Folder watching respects the ignores configured in Syncthing".  Could you point me to the place in the code where these ignores are applied?  

As far as I can tell almost all files that have been modified on my laptop during the last hour are actually ignored by .stignore, and this is confirmed by the "global changes" log that shows very few changes are actually propagated. Despite this the folders are scanned very often, far more often than any files are transferred.  If the .stignore patterns are applied *after* the notification logic in DirectoryWatcher.OnPathChanged finds a common prefix for all changed files, then it would explain what I'm seeing here:  A few ignored files are modified, the file watcher tells SyncTrayzor, we don't know yet that those files are ignored, their only common prefix turns out to be [AppData], and [AppData] isn't ignored so the entire folder is scanned.  But that is just speculation so far.

**Comment by canton7** [comment]
Looks like that bit of the README is out of date: Syncthing stopped telling me its ignores, so I stopped being able to apple them.

**Comment by JonasKvarnstrom** [comment]
That explains a lot!  Thanks for the info about adding notification support directly to Syncthing.  I'll wait and see what happens there.

Meanwhile I've moved a couple of larger folders (in terms of the number of files they contain) to a different location.  Since they are no longer nested within the main home folder I can sync them separately, which seems to save a lot of time. Then I created a link in the home folder and ignored the link in .stignore.  

I'll have to test this experiment for a while longer to see how it works, but given the circumstances this issue should probably be closed.

Thanks again for the very quick answers!
